### ISSUE TITLE
Issue with high contention on a single endpoint / resource

### ISSUE THREAD

**Issue by pbhogan** [jiangwenyuan/nuster#74]
I was evaluating nuster for a project and came across a pretty serious flaw. 

Let's say you set up 100 connections to make repeated requests from a single endpoint as fast as they possibly can—essentially simulating high traffic. And then have that endpoint cached by nuster with a TTL of 5 seconds.

First, unless the initial connections are spaced out significantly, all (or many?) of them get a cache miss at the same time and are passed through to the backend. Even if they are staggered, as soon as the cache TTL expires, again all (or many) get a cache miss and are passed on the backend. 

The result is the backend gets slammed with massive bursts of requests. Preventing this is in large part the point of using a cache. It appears that until at least one backend connection returns a response, and the cache is repopulated, all other cache misses are also let through.

What should happen is, when there is a cache miss, only a single connection is allowed to the backend to get new content while all other connections wait on that single response.

**Comment by jiangwenyuan** [comment]
@pbhogan Thanks, you are right.

I'm planning to add a new directive `wait TIME`, to solve this kind of issue. If there's no `wait` specified, then there's no wait happens, all requests are passed to backend. if there's `wait` specified, then first request is passed to backend and others will wait for `TIME` seconds(forever if TIME is 0).

**Comment by packeteer** [comment]
I assume in the case of identical POSTs, it will wait till the first
transaction is completed ?

**Comment by jiangwenyuan** [comment]
@packeteer No, it does not wait.

**Comment by pbhogan** [comment]
You might also consider taking the approach varnish does, which is to optionally allow a grace period during which stale data is served to other connections while that first connection is fetching from the backend.

They extend that to allowing serving stale data for a grace period while the backend is down too, which can be useful for deploys where read-only data is basically zero-downtime as a result.

It would also be neat to be able to chain these, so it'll serve stale data for some time limit, then fall back to queueing and waiting for fresh data for another time limit, and then finally either a backend fetch or failure.

And of course this would only really be relevant for GET/HEAD requests, although you can do others in varnish if you wanted to. But that seems like a weird, extremely niche use case.

**Comment by jiangwenyuan** [comment]
Thanks, and I'm also planning to add support for background update.

**Comment by jiangwenyuan** [comment]
Added `wait WAIT`

**Comment by uasan** [comment]
There is a similar request for HAProxy
https://github.com/haproxy/haproxy/issues/214#issuecomment-524842217

**Comment by jiangwenyuan** [comment]
@uasan Thanks, `cache lock` has been implemented by `wait`, and `Use stale cache while revalidate` is already done and I expect to push to master in a day or two after doing some test.

**Comment by jiangwenyuan** [comment]
@pbhogan @uasan Added `use-stale on|off|TIME`.

Now nuster supports `wait on|off|TIME`: one request send to backend to update cache and others wait`, `use-stale on|off|TIME`: use stale cache while updating cache or for TIME seconds when backend is down.
